py/torch_tensorrt/dynamo/converter_registry.py

```diff
@@ -0,0 +1,266 @@
+    check_args: Optional[Callable[[NodeKinds], bool]] = None,
+) -> Callable[[Any], Any]:
+    """Decorator for Dynamo TensorRT Converter
+
+    Registers the decorated function in the DYNAMO_ATEN_CONVERTERS registry
+
+    Args:
+        key: Node target for which the converter is implemented for
+            (for example, torch.ops.add.Tensor)
+        enabled: Whether the converter should be enabled/cached or not
+        check_args: Function which evaluates whether a node is valid for conversion
+            by the decorated converter. See ConverterSupport for more details.
+            Defaults to None, implying the check_args function is always true -
+            this means all nodes of "key" kind can be supported by this converter
+    Returns:
+        The converter being decorated
+    """
+
+    def register_converter(converter):
+        """Helper function to register the converter, then return it"""
+        assert callable(converter), "Converter function must be callable"
+
+        # If no check_args function is specified, use the default function - always return true
+        if check_args is None:
+            converter_support = ConverterSupport(converter_implementation=converter)
+        else:
+            assert callable(check_args), "Argument checking function must be callable"
+            converter_support = ConverterSupport(
+                converter_implementation=converter, check_args=check_args
+            )
+
+        # If a converter for this operator already exists, append the new converter to the list
+        # Otherwise, start a new list
+        if key in DYNAMO_ATEN_CONVERTERS:
+            DYNAMO_ATEN_CONVERTERS[key].append(converter_support)
+        else:
+            DYNAMO_ATEN_CONVERTERS[key] = [converter_support]
+
+        return converter
+
+    def disable_converter(converter):
+        return converter
+
+    # Select whether to cache/enable the converter
+    if enabled:
+        return register_converter
+    else:
+        return disable_converter
+
+
+class ConverterRegistry:
+    """Registry for storing multiple converter dictionaries
+
+    Capable of storing dictionaries with the following signature:
+    Dict[Target, Union[Callable, Sequence[ConverterSupport]]]
+
+    Also able to validate converter implementations against user-provided
+    argument-checking functions
+
+    Args:
+        registries: List of dictionaries representing converter registries.
+            The order of the provided dictionaries is the order in which they
+            will be traversed. This is only significant when using non-validated
+            methods.
+    """
+
+    def __init__(
+        self,
+        registries: Sequence[Dict[Target, Union[Callable, Sequence[ConverterSupport]]]],
+    ):
+        # Copy reference to each dictionary object into attribute list
+        self.registries = [registry for registry in registries]
+        self.validate_invariants()
+
+    def validate_invariants(self):
```

Comment:
Can we throw a warning if there is a collision in converter registration? We should have semantics for shadowing (i.e. probably dynamo converters shadows fx converters) and this should take into account capabilities, i.e. `aten::max_pool(dyn_only) != aten::max_pool(static_only)`

Reply:
For the current implementation, the distinction of `dynamic` and `static` is not yet present. We also need to consider multiple layers of shadowing. For example, there can be (and are, in #2070), FX and Dynamo converter registry entries with identical conversion capabilities, where one converter is simply more updated, etc.

If we are considering collisions within a single registry, I agree that it makes sense to differentiate on capabilities. I am wondering how we might combine this with the notion of a `check_args` function? For example, consider two converters for the same aten operation, like `torch.ops.aten._to_copy`. One might handle `device` casts, while the other handles `dtype` casts. What might be a good design to differentiate these from actual converter collisions?